web/elm/src/Build/Build.elm

```diff
@@ -199,206 +224,249 @@ getUpdateMessage model =
-                ( model, [] )
+                ( model, effects )
 
         BuildHistoryFetched (Ok history) ->
-            handleHistoryFetched history model
+            handleHistoryFetched history ( model, effects )
 
         BuildJobDetailsFetched (Ok job) ->
-            handleBuildJobFetched job model
+            handleBuildJobFetched job ( model, effects )
 
         BuildJobDetailsFetched (Err err) ->
             flip always (Debug.log "failed to fetch build job details" err) <|
-                ( model, [] )
+                ( model, effects )
 
         _ ->
-            ( model, [] )
+            ( model, effects )
 
 
-update : Msg -> Model -> ( Model, List Effect )
-update msg model =
-    let
-        ( newTopBar, topBarEffects ) =
-            TopBar.update (fromBuildMessage msg) model.topBar
+handleDelivery : Delivery -> ( Model, List Effect ) -> ( Model, List Effect )
+handleDelivery delivery ( model, effects ) =
+    case delivery of
+        KeyDown keycode ->
+            handleKeyPressed keycode ( model, effects )
 
-        ( newModel, dashboardEffects ) =
-            updateWithoutTopBar msg model
-    in
-    ( { newModel | topBar = newTopBar }
-    , topBarEffects ++ dashboardEffects
-    )
+        KeyUp keycode ->
+            if keycode == Keycodes.shift then
+                ( { model | shiftDown = False }, effects )
+
+            else
+                case Char.fromCode keycode of
+                    'T' ->
+                        ( { model | previousTriggerBuildByKey = False }, effects )
 
+                    _ ->
+                        ( model, effects )
 
-updateWithoutTopBar : Msg -> Model -> ( Model, List Effect )
-updateWithoutTopBar action model =
-    case action of
+        ClockTicked OneSecond time ->
+            let
+                newModel =
+                    { model
+                        | now = Just time
+                        , hoveredCounter = model.hoveredCounter + 1
+                    }
+            in
+            updateOutput
+                (Build.Output.Output.handleStepTreeMsg <|
+                    StepTree.updateTooltip newModel
+                )
+                ( newModel, effects )
+
+        ScrolledToBottom atBottom ->
+            ( { model | autoScroll = atBottom }, effects )
+
+        EventsReceived envelopes ->
+            envelopes
+                |> Build.Output.Output.handleEnvelopes
+                |> flip updateOutput
+                    ( model
+                    , case getScrollBehavior model of
+                        ScrollWindow ->
+                            effects ++ [ Effects.Scroll Effects.ToBottom ]
+
+                        NoScroll ->
+                            effects
+                    )
+
+        ElementVisible id isVisible ->
+            let
+                lastBuildId =
+                    model.history
+                        |> List.reverse
```

Comment:
I think List.Extras has a `last`, which is clearer and probably more efficient than `reverse >> head`, but that's just a nitpick.